src/routing/kad-dht.md

```diff
@@ -0,0 +1,719 @@
+identifier.
+
+DHT Clients MUST NOT advertise support for the libp2p Kademlia
+protocol identifier nor offer the libp2p Kademlia protocol
+identifier for incoming streams.
+
+## Kademlia Keyspace
+
+<!-- TODO: add LaTeX or MathML support and fix below paragraph -->
+Kademlia [`[0]`](#bibliography) operates on a binary keyspace defined as $\lbrace 0,1 \rbrace^m$. In
+particular, the IPFS Kademlia DHT uses a keyspace of length $m=256$, containing
+all bitstrings of 256 bits. The distance between any pair of keys is defined as
+the bitwise XOR of the two keys, resulting in a new key representing the
+distance between the two keys. This keyspace is used for indexing both nodes
+and content.
+
+The Kademlia node identifier is derived from the libp2p node's [Peer
+ID](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md). The
+Kademlia node identifier is computed as the digest of the SHA2-256 hash
+function of the binary representation of the Peer ID. The Kademlia identifier
+is a 256-bit number, which is used as the node's identifier in the Kademlia
+keyspace.
+
+Example:
+
+```sh
+PeerID b58 representation: 12D3KooWKudojFn6pff7Kah2Mkem3jtFfcntpG9X3QBNiggsYxK2
+PeerID CID representation: k51qzi5uqu5djx47o56x8r9lvy85co0sdf1yfbzxlukdq4irr8ssn3o7dpfasp
+PeerID hex representation: 0024080112209e3b433cbd31c2b8a6ebbdca998bd0f4c2141c9c9af5422e976051b1e63af14d
+Kademlia identifier (hex): e43d28f0996557c0d5571d75c62a57a59d7ac1d30a51ecedcdb9d5e4afa56100
+```
+
+## Routing Table
+
+The Kademlia Routing Table maintains contact information about other DHT
+Servers in the network. It has knowledge about all nearby nodes and
+progressively fewer nodes as the XOR distance increases. This structure allows
+efficient and rapid navigation of the network during lookups.
+
+### Bucket Size
+
+The Routing Table MUST contain information about at least `k` DHT Servers whose
+Kademlia Identifier shares a common prefix of length `l` with the local node,
+for every `l` in `[0, 255]`, provided such nodes exist. The set of `k` peers
+sharing a common prefix of length `l` with the local node is called the
+_bucket_ `l`.
+
+In practice, buckets with smaller indices will typically be full, as many nodes
+in the network share shorter prefix lengths with the local node. Conversely,
+buckets beyond a certain index usually remain empty, since it's statistically
+unlikely that any node will have an identifier sharing a very long common
+prefix with the local node. For more information see [bucket population
+measurements](https://github.com/probe-lab/network-measurements/blob/master/results/rfm19-dht-routing-table-health.md#peers-distribution-in-the-k-buckets).
+
+The IPFS Kademlia DHT uses a bucket size of `k = 20`. This corresponds to the
+`k` value as defined in the original Kademlia paper [`[0]`](#bibliography). The `k` value is also
+used as a replication factor and defines how many peers are returned to a
+lookup request.
```

Comment:
This paragraph seems specific to IPFS, is the rest of this section?

Reply:
https://github.com/ipfs/specs/pull/497#discussion_r2016997925